rfcs/asynchronous_setup.md

```diff
@@ -0,0 +1,75 @@
+# RFC 32: Asynchronous setup
+
+## Summary
+
+Extend testharness.js with a new API that makes it easier for test authors to
+defer subtest execution until the completion of some asynchronous operation.
+
+## Details
+
+Many testharness.js tests in WPT need to perform some asynchronous operation
+before subtests may be executed. For instance:
+
+- WebIDL tests need to fetch data (IDL files)
```

Comment:
At least right now, these use all sync tests after the async setup. Is it desirable to change those all to promise tests?

Reply:
No, we shouldn't do that, which suggests we have to make sync tests work at least if defined after async setup is done.

Reply:
Currently, "harness complete" means "no more tests." If we want to support that pattern, there will have to be some gap between "setup complete" and "harness complete." I mentioned this a little earlier, but it doesn't seem like there's an intuitive (i.e. non-error-prone) way to offer that (my only idea was to defer "harness complete" until the next event loop task).

Tests could do this themselves using our latest design with code like:

```js
var my_setup = some_operation();
my_setup.then(() => define_my_sync_tests);
promise_setup(my_setup);
```

Which is certainly not ergonomic, but if we're trying to support the IDL tests specifically, that may be acceptable since this logic is abstracted into the `idl_test` helper function.

Reply:
So do we want both these to be true?

* Sync tests should not run before the setup promise fulfills
* The fulfillment of the setup promise signals the end of the test

It seems hard to do both.

Reply:
> It seems hard to do both.

Agreed. I came to the same conclusion through a slightly different rationale above, and I suggested an alteration that would have synchronous tests always execute immediately. Although no one commented on that suggestion, I think your response here is sufficient indication that the proposal should be updated, So I've done that.

Reply:
I don't think that "The fulfillment of the setup promise signals the end of the test" should be true. Is some variation of actually currently true? I've never observed that to be the case, it seems like it's never too late to define additional subtests, which will then delay completion of the overall test.

Reply:
> Is some variation of actually currently true?

If you do something like this:

```
<script>
test(() => {}, 'foo');
onload = () => { test(() => {}, 'bar'); };
</script>
```
then only the 'foo' test is created. However, if I remove the 'foo' test, then the 'bar' test is created. `explicit_done` in `setup` can be used to override the heuristic.

Reply:
Defining tests in the `promise_setup` promise chain seems weird to me. Kinda like creating tests in the `setup(() => { ... })` function. Earlier it was suggested that creating `test` or `async_test` before `promise_setup` has fulfilled should not be allowed.

I don't understand what's bad about transitioning to `COMPLETE` in a task after `promise_setup` has fulfilled (using `setTimeout( ..., 0)`), and return a thenable from `promise_setup`. Or maybe if `onload` hasn't happened yet, wait for that. It's arbitrary, sure, but so is the current `onload` handling, and as shown in my previous example, it can miss tests. But testharness.js optimizes for test writing ergonomics.

Reply:
Alright, so this is about the point at which the (single) `Tests` object transitions to `COMPLETE`, which of course has to happen at some point. Glad there isn't some other way that defining new subtests stops working that I'd missed :)

Since using `test` or `async_test` during async setup can't be allowed to work and we'd stop waiting for tests very soon after, there will realistically only be one pattern that will work which almost all tests will use, so we should make that a nice one.

`promise_setup(...).then(() => { test(...) })` ought to work, but if for some other reason the code is already in an async function, then `await promise_setup(...); test(...)` should also. I don't know if those are precisely the same in terms of microtask scheduling, probably they are.

I wonder if there isn't a pattern for promise-returning functions one could follow here? For things that fire events it's very straightforward to do some stuff after any listeners have been invoked, is there no equivalent for promises?